src/lib/transition_frontier/snark_pool_refcount.ml

```diff
@@ -12,18 +13,57 @@ module type Inputs_intf = sig
      and type staged_ledger := Staged_ledger.t
 end
 
-module type S = Transition_frontier_extension_intf
-
 module Make (Inputs : Inputs_intf) :
-  S
+  Transition_frontier_extension_intf
   with type transition_frontier_breadcrumb := Inputs.Breadcrumb.t
-   and type input := unit = struct
+   and type input := unit
+   and type view := int Inputs.Transaction_snark_work.Statement.Table.t =
+struct
   module Work = Inputs.Transaction_snark_work.Statement
 
-  type t = {ref_table: int Work.Table.t}
+  type t = int Work.Table.t
+
+  let get_work (breadcrumb : Inputs.Breadcrumb.t) : Work.t Sequence.t =
+    let ledger = Inputs.Breadcrumb.staged_ledger breadcrumb in
+    let scan_state = Inputs.Staged_ledger.scan_state ledger in
+    let work_to_do =
+      Inputs.Staged_ledger.Scan_state.all_work_to_do scan_state
+    in
+    Or_error.ok_exn work_to_do
+
+  let add_breadcrumb_to_ref_table table breadcrumb =
+    Sequence.iter (get_work breadcrumb) ~f:(fun work ->
+        Work.Table.change table work ~f:(function
+          | Some count -> Some (count + 1)
+          | None -> Some 1 ) )
+
+  let remove_breadcrumb_from_ref_table table breadcrumb =
+    Sequence.fold (get_work breadcrumb) ~init:false ~f:(fun acc work ->
+        match Work.Table.find table work with
+        | Some 1 ->
+            Work.Table.remove table work ;
+            acc || true
+        | Some v ->
+            Work.Table.set table ~key:work ~data:(v - 1) ;
+            acc || false
+        | None -> failwith "Removed a breadcrumb we didn't know about" )
 
-  let create () = {ref_table= Work.Table.create ()}
+  let create () = Work.Table.create ()
 
-  (* TODO: implement diff-handling functionality *)
-  let handle_diff (_t : t) _diff = ()
+  let handle_diff t diff =
+    match (diff : Inputs.Breadcrumb.t Transition_frontier_diff.t) with
+    | New_breadcrumb breadcrumb ->
+        add_breadcrumb_to_ref_table t breadcrumb ;
+        None
+    | New_best_tip {old_root; new_root; new_best_tip; garbage; _} ->
+        add_breadcrumb_to_ref_table t new_best_tip ;
+        let all_garbage =
+          if phys_equal old_root new_root then garbage else old_root :: garbage
+        in
+        let removed : bool =
+          List.fold ~init:false
+            ~f:(fun acc bc -> acc || remove_breadcrumb_from_ref_table t bc)
+            all_garbage
+        in
+        if removed then Some t else None
```

Comment:
Is the intent that the snark pool iterates over every snark it contains, checking whether they have >0 references whenever this view updates? It'd be more efficient for the view to be a list of now-garbage works instead. But unless I'm missing something the pool also needs to know whether a snark is valid/relevant when it first comes in, in which case the view should update any time a work's refcount is added or removed from the table, not just when the best tip changes.

Reply:
Ah right, it's a good point that the view needs to update when new work gets added to the table, thanks.

As far as the "now-garbage" list is concerned, that was closer to the original design, but @bkase and I were wondering if it might be simpler to periodically (not necessarily every time the view updates) go through the snark_pool and remove anything that isn't referenced.

Reply:
It means we have one less part of the system that has to stay up-to-date with every event as it flows through (desirable imo)